src/index.js

```diff
@@ -1,99 +1,83 @@
-
-      static wrappedComponent = WrappedComponent
-
-      static defaultProps = {
-        serverRendered: false
-      }
-
-      constructor(props, context) {
-        super(props, context);
-        if (!this.isOnServer) {
-          this.state = {
-            dependenciesResolved: this.props.serverRendered,
-            dependenciesFailed: false
-          }
-        }
-        this.onDependencyResolution = this.onDependencyResolution.bind(this);
-        this.onDependencyFailure = this.onDependencyFailure.bind(this);
-      }
-
-      static arrayPropertyIsValid(component, propertyName) {
-        return Array.isArray(component[propertyName]) && component[propertyName].length > 0;
-      }
-
-      get isOnServer() {
-        return !(typeof window !== 'undefined' && window.document);
-      }
-
-      static componentHasDependencies(component) {
-        return this.arrayPropertyIsValid(component, 'dependencies')
-      }
-
-      static componentDependencies(component) {
-        return component.dependencies.map(dependency => {
-          return dependency()
-        })
-      }
-
-      onDependencyResolution(results=[]) {
-        this.setState({
-          dependenciesResolved: true
-        })
+    const DependenciesResolver = props => {
+      const [dependenciesResolved, setDependenciesResolved] = useState(false);
+      const [dependenciesFailed, setDependenciesFailed] = useState(false);
+      const [reason, setReason] = useState(null);
+
+      const arrayPropertyIsValid = (component, propertyName) => {
+        return (
+          Array.isArray(component[propertyName]) &&
+          component[propertyName].length > 0
+        );
+      };
+
+      const componentHasDependencies = component =>
+        arrayPropertyIsValid(component, 'dependencies');
+
+      const componentDependencies = component =>
+        component.dependencies.map(dependency => dependency(props), this);
+
+      const dependencies = component =>
+        componentHasDependencies(component)
+          ? componentDependencies(component)
+          : [Promise.resolve('no component dependencies')];
+
+      const onDependencyResolution = (results = []) => {
+        setDependenciesResolved(true);
         if (WrappedComponent.onDependencyResolution) {
-          WrappedComponent.onDependencyResolution(results)
+          WrappedComponent.onDependencyResolution(results);
         }
-      }
-
-      onDependencyFailure(reason) {
-        this.setState({
-          dependenciesFailed: true
-        })
-        if (WrappedComponent.onDependencyFailure) {
-          WrappedComponent.onDependencyFailure(reason)
+      };
+
+      const onDependencyFailure = reason => {
+        const { isCanceled } = reason;
+
+        if (!isCanceled) {
+          // If we have cancelled the promise when unmounting the component, don't
+          // attempt to update any state values.
+          setDependenciesFailed(true);
```

Comment:
These state updates were causing the memory leaks before. Now if we unmount, always reject the promise so we call onDependencyFailure and if the reason is cancellation (which happens when we unmount), skip all additional callbacks and state update actions.